core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala

```diff
@@ -7,186 +7,170 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
+    case Unregistered        ⇒ context become waitingToTerminate(outstanding)
+    case Terminated(`queue`) ⇒ context become waitingToTerminate(outstanding)
 
-        case m ⇒ Left(s"Unmatched Result '${descriptionOf(m)}' from the backend service, update your ResultChecker if you want to prevent it from being treated as an error.")
+    //work completed on way or another, just waiting for the Unregister ack
+    case WorkFinished        ⇒ context become unregistering
+  }
 
-      }: ResultChecker)).andThen[Unit] {
+  def becomeUnregistering(): Unit = {
+    queue ! Unregister(self)
+    context become unregistering
+  }
 
-        case Right(result) ⇒
-          outstanding.success(result)
-          workFinished(isRetiring)
+  //onRouteeFailure is what gets called if while waiting for a Routee response, the Routee dies.
+  def routeeResponse(outstanding: Outstanding, onRouteeFailure: () ⇒ Unit): Receive = {
+    case Terminated(`routee`) ⇒ {
+      outstanding.fail(WorkFailed(s"due ${routee.path} is terminated"))
+      onRouteeFailure()
+    }
+    case x if sender() == routee ⇒ {
+      val result: Either[String, Any] = resultChecker.applyOrElse(x, failedResultMatch)
+      result match {
+        case Right(res) ⇒
+          outstanding.success(res)
+          self ! WorkFinished
 
         case Left(e) ⇒
           log.warning(s"Error $e returned by routee in regards to running work $outstanding")
-          retryOrAbandon(outstanding, isRetiring, e)
+          retryOrAbandon(outstanding, e)
       }
+    }
+    case RouteeTimeout ⇒
+      log.warning(s"Routee ${routee.path} timed out after ${outstanding.work.settings.timeout} work ${outstanding.work.messageToDelegatee} abandoned")
+      outstanding.timeout()
+      self ! WorkFinished
 
-    ({
-      case RouteeTimeout ⇒
-        log.warning(s"Routee ${routee.path} timed out after ${outstanding.work.settings.timeout} work ${outstanding.work.messageToDelegatee} abandoned")
-        outstanding.timeout()
-        workFinished(isRetiring)
-
-      case w: Work ⇒ sender ! Rejected(w, "busy") //just in case
-
-    }: Receive) orElse handleResult
   }
 
-  def workFinished(isRetiring: Boolean): Unit = {
-    if (isRetiring) {
-      finish()
-    } else {
-      askMoreWork()
-    }
+  def failedResultMatch(x: Any): Either[String, Any] = {
+    Left(s"Unmatched Result '${descriptionOf(x)}' from the backend service, update your ResultChecker if you want to prevent it from being treated as an error.")
   }
 
-  private def retryOrAbandon(
-    outstanding: Outstanding,
-    isRetiring:  Boolean,
-    error:       Any
-  ): Unit = {
+  private def retryOrAbandon(outstanding: Outstanding, error: Any): Unit = {
     outstanding.cancel()
-    //why do we fail if there is a delayBeforeNextWork? Is this because of the subsequent 'sendWorkToDelegatee' call?
+    //why do we fail if there is a delayBeforeNextWork? Is this because of the subsequent 'sendWorkToRoutee' call?
```

Comment:
mainly because if there is hold, you shouldn't send work to routee, and it's might be too late to try again after the delay. So fail immediately might be the safest option.